### ISSUE TITLE
Binary variable and big_M seems to give invalid solutions for not so large M

### ISSUE THREAD

**Issue by krchiper** [Specy/microlp#7]
I am not sure if this is a bug or a consequence stated by the warning big_M comes with.

```
min x - b
subject to
(1)    x >= y
(2)    -M * b + x <= 10
(3)    -M * b + x >= -M + 10
where
    let y = 1        
    let M = 1000000  // <-- if changed to 100000, the solution becomes correct
define
    x as NonNegativeReal
    b as Boolean
```
Which gives the solution `x = 1` and `b = 1`, this violates constraint 3
```
(3)  -M * b + x >= -M + 10  =>
     -1000000 * 1 + 1 >= -1000000 + 10 =>
     -999999 >= -999990 
```
Reducing `M` by an order of magnitude, fixes the example and a correct solution is found.

link to [example](https://rooc.specy.app/projects/share?project=N4IgbgpgTgzglgewHYgFwEYA0IoFckAucAthGiAPQBUAOkgMbIwEAEAZnADYQwsC8LAOIBRACoB9AGIBJADLCAygAoAlHTpQIAczjNoS4HRbGWSAIalULAOQwAjlALXMRkwBMe9KHAAORZFbWAMJmnPS4nGYEECwEABYx9rhmmixQCAisCGwsZqa4xABG0M6uxj4pFhDRsFYA2nXW2eJIBc4sAArexHBEkAB0AHIFxVAAumMuSCZp1bhQSDBWXSS9cAPDRdBTM-ShnFZKrcQq-AB8LIbTMyaaBPPTwAQAnj4QVjQgm6OfmCxgoVw7xYAFkonF+vZHEcCv0AZwgSoAL5lFgopBItRIOiMRYEMyEGAGModaQACVCbCswFir2Bn2+0F+-0BwLB8X6pJYFBYACY0XRMXQqBQQNg8IQSBAAGq6OCFbhoNihGAQbDmUjkaLMFiFOBacQgsUgDwwLy+fwoVAgY1eCBRCBuACCBDQ6AA7ABmACsvM97oADOhee6AJye7C4HxuB3O10YL2+z0ADm93vQ6ADyewHG4MDQdUmIFx0UI5B60wAHiwALS6ugwXCFABWEHorAICDoSnQpxM1bOAme3d5feMNZBLCoupYAGoWNWADwCTPdz1j2uT6eFOcLliDze71dIADuCU0qO4rGe-BY6Ev1VBt8zAdfr+MFB5i5rdbgOXocQElojqxAgd5vq+fzxIkCAIpauptggpC8IwUCaO2dAeBwSAQKi1ZmLwgzIIM2hROsEAAEr2pwqI7gRLAAEIZNwBLGj4vg8AWoDsW8aABtgCBvFayqcKqSKYNxHFugJQlKiqEDiZJvGoLyMkQMJ8mKSAPFkKgEYgIJ6loAQeAKRJ2lSRgWAGbJqAmUCSJFnAbjkDAgGaCASJAA) on rooc

**Comment by kwcarlsen** [comment]
The constraints are reduced from another problem to give a minimal failing test case. 

It comes from optimization of electricity usage with a solar installation to get the lowest utility bill. If you use more than 4.000 kWh electricity, you get a tariff reduction of everything above.

```
MIN Z = <objective> - [ if x <= 4000 then 0 else x - 4000 ]
```
This can be re-written using a binary variable and big_m

**Comment by Specy** [comment]
No I mean, the number inside parenthesis in each constraint is actually part of the constraint, it's not a comment or similar. 

```
(2)    x >= y
(3)    -M * b + x <= 10
```
Is 
```
2x >= y
3-M*b+X<=10
```

Is this your intended behaviour? Or just did it as an example?

**Comment by krchiper** [comment]
You are right, this wasn't my intention :smile:. I tried to assign numbers to the constraints so I was able to reference them later. It's just a coincidence it didn't break my example.
